linkml/generators/pydanticgen/build.py

```diff
@@ -0,0 +1,44 @@
+from typing import List, Optional, Type, Union
+
+from pydantic import BaseModel
+
+from linkml.generators.pydanticgen.template import Import, Imports
+
+
+class BuildResult(BaseModel):
+    """
+    The result of any build phase for any linkML object
+
+    BuildResults are merged in the serialization process, and are used
+    to keep track of not only the particular representation
+    of the thing in question, but any "side effects" that need to happen
+    elsewhere in the generation process (like adding imports, injecting classes, etc.)
+    """
+
+    # FIXME: PLACEHOLDER TYPES PENDING MERGE OF OTHER PULL REQUESTS
+    imports: Optional[Union[List[Import], Imports]] = None
+    injected_classes: Optional[List[Union[str, Type]]] = None
+
+    def __add__(self, other: "BuildResult"):
+        self_copy = self.copy()
+        if other.imports:
+            if self.imports is not None:
+                self_copy.imports += other.imports
+            else:
+                self_copy.imports = other.imports
+        if other.injected_classes:
+            self_copy.injected_classes = self_copy.injected_classes.extend(other.injected_classes)
+        return self_copy
+
+
+class SlotResult(BuildResult):
+    annotation: str
+    """The type annotation used in the generated model"""
+    field_extras: Optional[dict] = None
+    """Additional metadata for this slot to be held in the Field object"""
+
+    def __add__(self, other: "SlotResult"):
+        res = super(SlotResult, self).__add__(other)
+        # Replace with other's annotation
+        res.annotation = other.annotation
+        return res
```

Comment:
The replacement of `annotation` when using `+=` is a bit less than intuitive to me. It looks like it is used in only two places above: lines 361 and 382 of `array.py`. I get that the imports and injected classes are additive and that makes for cleaner code in `self._labeled_dimension` and elsewhere, but when seeing `+=`, I had the incorrect intuition that `annotation` was also additive, like it is a conlist, and `+=` would add a constraint to the conlist. Instead, it is used like `res += function(arg, dtype=res.annotation)` which nests/wraps the previous annotation and adds any relevant imports and injected classes needed by `function`. 

It makes sense now after tracing the code, and it is nicely compact, but I wonder if it could and should be refactored to be more intuitive, at the cost of more lines of code. Instead of using `+=`, we could explicitly create a new `SlotResult`. For example, instead of lines 381-382:

```python
for dim in reversed(array.dimensions):
    res += self._labeled_dimension(dim, dtype=res.annotation)
```
(note: this might add `_ConListImports` N times to the resulting `SlotResult`?)

Use:
```python
ann = res.annotation
for dim in reversed(array.dimensions):
    ann = self._labeled_dimension(dim, ann)
res = SlotResult(
    annotation=ann,
    injected_classes=res.injected_classes,
    imports=res.imports.copy() + _ConListImports,
)
```

Now that I write it out, I see that this is pretty verbose. I'm not sure if it is better, but it is more clear to me what's going on. I've just been scarred recently by unintuitive `+=` behavior elsewhere (see below)...

Note that on lines 317-322, we have something similar:
```python
return SlotResult(
    annotation=self._list_of_lists(array.minimum_number_dimensions - 1, self.any_shape().annotation),
    injected_classes=_AnyShapeArrayInjects,
    imports=_AnyShapeArrayImports,
)
```

Fun exercise:
```python
a = np.array([1, 2, 3])
b = a[1:3]
b += 2
print(b)  # [4 5]
print(a) # [1 4 5] !!!
# this happens because numpy slices are views of the array and `+=` acts in place
# compare that with:
a = np.array([1, 2, 3])
b = a[1:3]
b = b + 2
print(b)  # [4 5]
print(a) # [1 2 3]
# one might think that b += 2 is the same as b = b + 2, but no, b + 2 created a copy
```

Reply:
Fair point!

So the expectation with these BuildResults is that they are going to be part of a potentially unknown processing chain in the rest of the generator. Each "build stage" should receive the existing BuildResult and return a modified one. Replacing the annotation is from the expectation that between stages, the build method wouldn't know about the internal structure of the generated annotation. This is sort of the cost of doing code generation - at some point it becomes a string unless you build a full-on ast. This first draft is not intended to be final, but just a starting place, so that can of course be varied (eg. it probably would be good to carry the `SlotDefinition` along with the result to, etc.). 

One thing's for sure - this needs to be documented better than i have it now, regardless of the behavior.

> when seeing +=, I had the incorrect intuition that annotation was also additive, like it is a conlist, and += would add a constraint to the conlist.

Also a fair assumption - this would be awesome, but super tricky to implement in a general way. I think the intuition for this will be clearer when/if used in the rest of the generator - the annotation is specifically the string that is going to be rendered in the final model, and each of the build stages can't know the programmatic format of the annotation. 

The main thing that the results are for is to contain all the side effects that happen throughout the build process - specifically knowing which imports and etc. need to happen to make each of the different slots possible to use without needing to inject everything every time. So I think we should avoid doing a ton of modification of the build result in the middle of the method because that sort of kills the generality of that as a container - eg. we have some other metamodel addition later that adds a field to the buildresult, now we need to go through and modify all the places where we made custom addition logic instead of just updating the `__add__` method.

> scarred recently by unintuitive += behavior elsewhere

youch! yes that is why this method always returns a new object instead of modifying in place.

Reply:
swapped `__add__` to `merge()`